lib/semantic_puppet/version_range.rb

```diff
@@ -1,422 +1,729 @@
-      #
-      # @overload parse_inequality_expression('>=', expr)
-      #   {include:.parse_gte_expression}
-      #
-      # @param comp ['<', '<=', '>', '>='] an inequality operator
-      # @param expr [String] a "loose" version expression
-      # @return [VersionRange] a range covering all versions in the inequality
-      def parse_inequality_expression(comp, expr)
-        case comp
-        when '>'
-          parse_gt_expression(expr)
-        when '>='
-          parse_gte_expression(expr)
-        when '<'
-          parse_lt_expression(expr)
-        when '<='
-          parse_lte_expression(expr)
-        end
-      end
+      version = Version.new(major, minor, patch, Version.parse_prerelease(match_data[4]), Version.parse_build(match_data[5]))
+      return EqRange.new(version) unless tilde_or_caret
 
-      # Returns a range covering all versions greater than the given `expr`.
-      #
-      # @param expr [String] the version to be greater than
-      # @return [VersionRange] a range covering all versions greater than the
-      #         given `expr`
-      def parse_gt_expression(expr)
-        if expr =~ /^[^+]*-/
-          start = Version.parse("#{expr}.0")
+      MinMaxRange.new(GtEqRange.new(version), LtRange.new(Version.new(major, minor + 1, 0)))
+    end
+    private_class_method :allow_patch_updates
+
+    def self.allow_minor_updates(major, match_data)
+      return AllRange.SINGLETON unless major
+      minor = digit(match_data[2])
+      if minor
+        patch = digit(match_data[3])
+        if patch.nil?
+          MinMaxRange.new(GtEqRange.new(Version.new(major, minor, 0)), LtRange.new(Version.new(major + 1, 0, 0)))
         else
-          start = process_loose_expr(expr).last.send(:first_prerelease)
+          if match_data[4].nil?
+            MinMaxRange.new(GtEqRange.new(Version.new(major, minor, patch)), LtRange.new(Version.new(major + 1, 0, 0)))
+          else
+            MinMaxRange.new(
+              GtEqRange.new(
+                Version.new(major, minor, patch, Version.parse_prerelease(match_data[4]), Version.parse_build(match_data[5]))),
+              LtRange.new(Version.new(major + 1, 0, 0)))
+          end
         end
-
-        self.new(start, SemanticPuppet::Version::MAX)
+      else
+        MinMaxRange.new(GtEqRange.new(Version.new(major, 0, 0)), LtRange.new(Version.new(major + 1, 0, 0)))
       end
+    end
+    private_class_method :allow_minor_updates
 
-      # Returns a range covering all versions greater than or equal to the given
-      # `expr`.
-      #
-      # @param expr [String] the version to be greater than or equal to
-      # @return [VersionRange] a range covering all versions greater than or
-      #         equal to the given `expr`
-      def parse_gte_expression(expr)
-        if expr =~ /^[^+]*-/
-          start = Version.parse(expr)
+    def self.digit(str)
+      str.nil? || UPPER_X == str || LOWER_X == str || STAR == str ? nil : str.to_i
```

Comment:
Similar to previous comments, I would suggest wrapping the condition in a set of parentheses just to clarify precedence.